File: lore/mapas/ferramentas/static/js/geo.js
// Distância real sobre Uldun (grande círculo, haversine), do lado do navegador.
//
// Arquivo próprio desde a etapa 9 (2026-09-23): a régua (regua.js) mede com esta
// função, e o servidor mede com backend/geo.py (é com ela que a atração das estradas
// decide os 5 km; a tela da Estrada só MOSTRA os km que o servidor devolve, e não
// mede nada). Quem mais precisar medir no navegador usa esta, e não uma cópia.
// As duas implementações são conferidas uma contra a outra, nos mesmos pontos, por
// tests/test_haversine.py, que roda ESTE arquivo no node (não uma cópia colada no
// teste). Mudou uma, o teste acusa.
//
// O raio vem de fora (PARAMETROS_LEAFLET.raio_km, injetado de dados/coordenadas.json)
// e nunca é digitado aqui. O Math.min(1, ...) é o mesmo do servidor: perto do
// antípoda o arredondamento pode deixar `a` um fio acima de 1, e asin daria NaN.

function haversineKm(lat1, lon1, lat2, lon2, raioKm) {
  const rad = Math.PI / 180;
  const dLat = (lat2 - lat1) * rad;
  const dLon = (lon2 - lon1) * rad;
  const a = Math.sin(dLat / 2) ** 2
    + Math.cos(lat1 * rad) * Math.cos(lat2 * rad) * Math.sin(dLon / 2) ** 2;
  return 2 * raioKm * Math.asin(Math.min(1, Math.sqrt(a)));
}

File: lore/mapas/ferramentas/backend/geo.py
"""Distância real sobre Uldun (grande círculo), do lado do servidor.

É a fonte única da fórmula no Python: a atração automática das estradas
(`backend/estradas.py`) mede com ela, e `tests/test_haversine.py` a importa daqui em
vez de manter uma cópia própria (até a etapa 9 o teste tinha a sua).

No navegador a fórmula precisa existir de novo, em JavaScript, e mora num arquivo só:
`static/js/geo.js`, usado pela régua. As duas não podem ser uma só (uma roda no
servidor, a outra na página), então o que garante que dizem o mesmo é
`tests/test_haversine.py`, que roda o `geo.js` de verdade no node e compara com esta
função nos mesmos pontos, com um controle negativo que prova que a comparação pega
uma fórmula alterada.

O raio é o de Uldun (`dados/coordenadas.json`, `planeta.raio_km`), nunca o da Terra:
o planeta tem 1,25 vez o raio terrestre, e usar o número errado erraria toda
distância por 25%.
"""

import math

from . import coordenadas


def raio_km() -> float:
    return coordenadas.carregar_coordenadas()["planeta"]["raio_km"]


def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float,
                 raio: float | None = None) -> float:
    raio = raio if raio is not None else raio_km()
    f1, f2 = math.radians(lat1), math.radians(lat2)
    df = math.radians(lat2 - lat1)
    dl = math.radians(lon2 - lon1)
    a = math.sin(df / 2) ** 2 + math.cos(f1) * math.cos(f2) * math.sin(dl / 2) ** 2
    return 2 * raio * math.asin(min(1.0, math.sqrt(a)))
